plugins/notifications-backend/src/service/handlers.ts

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint-disable func-names */
+import { CatalogClient } from '@backstage/catalog-client';
+
+import { Knex } from 'knex';
+
+import { Components, Paths } from '../openapi';
+import {
+  DefaultMessageScope,
+  DefaultOrderBy,
+  DefaultOrderDirection,
+  DefaultPageNumber,
+  DefaultPageSize,
+} from './constants';
+import { ActionsInsert, dbValToBoolean, MessagesInsert } from './db';
+import {
+  MessageScopes,
+  NotificationsFilterRequest,
+  NotificationsOrderByDirections,
+  NotificationsOrderByFields,
+  NotificationsSortingRequest,
+} from './types';
+
+// createNotification
+// returns string id of created notification
+export async function createNotification(
+  dbClient: Knex<any, any>,
+  catalogClient: CatalogClient,
+  req: Paths.CreateNotification.RequestBody,
+): Promise<Paths.CreateNotification.Responses.$200> {
+  let isUser = false;
+
+  // validate users
+  if (Array.isArray(req.targetGroups) && req.targetGroups.length > 0) {
+    isUser = true;
+    const promises = req.targetGroups.map(group => {
+      return catalogClient.getEntityByRef(`group:${group}`).then(groupRef => {
+        if (!groupRef) {
+          throw new Error(`group '${group}' does not exist`);
+        }
+      });
+    });
+
+    await Promise.all(promises);
+  }
+
+  // validate groups
+  if (Array.isArray(req.targetUsers) && req.targetUsers.length > 0) {
+    isUser = true;
+    const promises = req.targetUsers.map(user => {
+      return catalogClient.getEntityByRef(`user:${user}`).then(userRef => {
+        if (!userRef) {
+          throw new Error(`user '${user}' does not exist`);
+        }
+      });
+    });
+
+    await Promise.all(promises);
+  }
+
+  // validate actions
+  if (Array.isArray(req.actions)) {
+    req.actions.forEach(action => {
+      if (!action.title || !action.url) {
+        throw new Error('Both action title and url are mandatory.');
+      }
+    });
+  }
+
+  const row: MessagesInsert = {
+    origin: req.origin,
+    title: req.title,
+    message: req.message,
+    topic: req.topic,
+    is_system: !isUser,
+  };
+
+  let messageId: string;
+
+  const ret = dbClient.transaction(trx => {
```

Comment:
The request handlers and database should be separated so that it's possible to overwrite the default functionality of the database. This is done in other plugins by creating interface for the store and then creating default implementation for that.

Reply:
I got your point. But do you have an example where more implementations are actually used?
Any unneccessary abstraction makes code harder to read and maintain. And shows to be wrong in the long-term (based on practical experience).

Reply:
then let's wait for real need. no point in building infrastructure and code for something not used. I agree with Marek that the code is much simpler this way.
if it is for other developers as well then they contribute based on their needs.

Reply:
Having a simple interface that can be overridden by the application developers is not that much extra work. Yeah, it's simpler but one should always remember that the users of each and every plugin is someone who is managing their own backstage instance and probably who want to have a possibility to modify the default functionality if they wish. Most of these use cases are hidden as people don't share problems until they hit one. Following the same architecture as other plugins should be something to work on.